File: examples/python/intent-output-p2pkh/IntentOutputP2PKH.runar.py
from runar import (
    StatefulSmartContract, ByteString, Bigint, Readonly, public,
)


class IntentOutputP2PKH(StatefulSmartContract):
    """Exercises the requireOutputP2PKH intent intrinsic.

    Asserts that output 0 of the spending transaction is a standard
    P2PKH output paying exactly bondAmount satoshis to bondPKH.

    Note: The properties bondPKH and the builtin requireOutputP2PKH are
    spelled in camelCase here (rather than the usual Python snake_case)
    because the all-caps PKH token does not round-trip through the
    naive snake_case <-> camelCase transform shared across the
    cross-tier Python parsers. Writing them in their canonical
    camelCase form avoids the lossy conversion and keeps the AST
    byte-identical to the Go-source ASTAfter Go parser lowercases the
    leading char of BondPKH/RequireOutputP2PKH.
    """

    bondPKH: Readonly[ByteString]
    bondAmount: Readonly[Bigint]
    count: Bigint

    def __init__(self, bondPKH: ByteString, bondAmount: Bigint, count: Bigint):
        super().__init__(bondPKH, bondAmount, count)
        self.bondPKH = bondPKH
        self.bondAmount = bondAmount
        self.count = count

    @public
    def payBond(self):
        requireOutputP2PKH(0, self.bondPKH, self.bondAmount)
        self.count = self.count + 1

File: examples/rust/intent-output-p2pkh/IntentOutputP2PKH.runar.rs
use runar::prelude::*;

/// IntentOutputP2PKH -- exercises the `requireOutputP2PKH` intent
/// intrinsic. Asserts that output 0 of the spending transaction is a
/// standard P2PKH output paying exactly `bondAmount` satoshis to
/// `bondPKH`.
///
/// Note: the property `bondPKH` and the builtin `requireOutputP2PKH`
/// are spelled in camelCase here (rather than the usual Rust
/// snake_case) because the all-caps `PKH` token does not round-trip
/// through the naive snake_case <-> camelCase transform shared across
/// the cross-tier Rust parsers. Writing them in their canonical
/// camelCase form avoids the lossy conversion and keeps the AST
/// byte-identical to the Go-source AST after the Go parser lowercases
/// the leading char of `BondPKH` / `RequireOutputP2PKH`.
#[runar::contract]
pub struct IntentOutputP2PKH {
    #[readonly]
    pub bondPKH: ByteString,
    #[readonly]
    pub bondAmount: Bigint,
    pub count: Bigint,
}

impl IntentOutputP2PKH {
    pub fn payBond(&mut self) {
        requireOutputP2PKH(0, self.bondPKH, self.bondAmount);
        self.count = self.count + 1;
    }
}
